cleanNodes.groovy

```diff
@@ -0,0 +1,65 @@
+#!/bin/env groovy
+/*  A system script which will gather node items
+    about any system its being run from and will
+    remove the workspace folders of each of the
+    folders it holds recursively based on whether
+    it have ever been started, when the last build
+    was kicked off, if it is currently building,
+    while setting the node offline.
+    partly derived from source: http://blog.ehrnhoefer.com/2016-06-14-jenkins-workspace-cleanup/
+ */
+
+import hudson.model.*;
+import hudson.matrix.*;
+import hudson.maven.*;
+import hudson.util.*;
+import jenkins.model.*;
+import hudson.FilePath.FileCallable;
+import hudson.slaves.OfflineCause;
+import hudson.node_monitors.*;
+
+Jenkins.instance.nodes.each { node ->
+    computer = node.toComputer()
+    if (computer.getChannel() == null) continue
+
+    rootPath = node.getRootPath()
+    size = DiskSpaceMonitor.DESCRIPTOR.get(computer).size
+    roundedSize = size / (1024 * 1024 * 1024) as int
+
+    println("node: " + node.getDisplayName() + ", free space: " + roundedSize + "GB")
+    if (roundedSize < 10) {
+        computer.setTemporarilyOffline(true, new hudson.slaves.OfflineCause.ByCLI("disk cleanup"))
+        Jenkins.instance.getAllItems(Job.class).each { item ->
+
+            // MavenModule is superfluous project returned by getAllItems()
+            if (!(item instanceof MatrixConfiguration || item instanceof MavenModule)) {
```

Comment:
Could I ask why we dont care about running Maven or Matrix jobs here? Or does isBuilding() just not work on them?

Reply:
Maven modules are items (they, too, have a unique config.xml) within a Maven project job and seem to be managed as items by jenkins just as jobs are even though they are inside the same workspace.  So if the Maven project job is cleaned first (including all already built modules), Jenkins will throw an exception when it attempts to find the maven module beneath it.